File: examples/julia/tasks/object_detection/object_detection.jl
using Bytez
using HTTP
using Images
using Colors
using FileIO
using Luxor

WORKING_DIR = dirname(@__FILE__)

client = Bytez.init("YOUR BYTEZ KEY HERE")

model = client.model("facebook/detr-resnet-50")

model.load()

img_url = "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Cat_November_2010-1a.jpg/1200px-Cat_November_2010-1a.jpg"

result = model.run(img_url)

box_objects = result["output"]

# Print each box object with additional information
for box_object in box_objects
	# Depending on the model, there may be additional properties returned
	println(box_object)

	score = box_object["score"]
	label = box_object["label"]
	box = box_object["box"]

	println(Dict("score" => score, "label" => label, "box" => box))
end

# Fetch image from URL
response = HTTP.get(img_url)
image_bytes = response.body

# Load image and convert to RGB format
image = load(IOBuffer(image_bytes))
image_rgb = colorview(RGB, image)
width, height = size(image_rgb)

# Save the loaded image temporarily as PNG for Luxor compatibility
temp_image_path = joinpath(WORKING_DIR, "temp_image.png")
save(temp_image_path, image_rgb)

# Read the dimensions of the saved temporary image
temp_image = load(temp_image_path)
temp_height, temp_width = size(temp_image)

# Visualization function using Luxor
function debug_image(src_img_path, box_objects, output_path, img_width, img_height)
	colors = [
		RGB(1.0, 0.0, 0.0),  # Red
		RGB(0.0, 1.0, 0.0),  # Green
		RGB(0.0, 0.0, 1.0),  # Blue
		RGB(1.0, 1.0, 0.0),   # Yellow
	]

	# Start a drawing with exact dimensions of the source image
	Drawing(img_width, img_height, output_path)
	origin(Point(0, 0))  # Keep the origin at the top-left corner

	# Load the PNG image and place it on the drawing canvas
	src_img = readpng(src_img_path)
	placeimage(src_img, 0, 0, centered = false)

	# Draw bounding boxes and labels
	for (index, box_object) in enumerate(box_objects)
		box = box_object["box"]
		score = box_object["score"]
		label = box_object["label"]
		color = colors[mod1(index, length(colors))]

		# Set color for the bounding box
		sethue(color)

		# Draw bounding box
		xmin, ymin, xmax, ymax = box["xmin"], box["ymin"], box["xmax"], box["ymax"]
		box_width, box_height = xmax - xmin, ymax - ymin
		box_position = Point(xmin, ymin)
		@layer begin
			rect(box_position, box_width, box_height, :stroke)
		end

		# Draw label and score above bounding box
		sethue("white")
		fontsize(12)
		text_position = Point(xmin, ymin - 10)
		text("$label ($(round(score, digits=2)))", text_position, halign = :left)
	end

	# Finish and save the drawing
	finish()
end

# Define output path and call the function with exact dimensions
output_path = joinpath(WORKING_DIR, "test_image.png")
debug_image(temp_image_path, box_objects, output_path, temp_width, temp_height)

File: examples/python/tasks/object_detection/object_detection.py
import os
import requests
from PIL import Image, ImageDraw, ImageFont
from bytez import Bytez
import io

WORKING_DIR = os.path.dirname(os.path.realpath(__file__))

client = Bytez("YOUR_BYTEZ_KEY_HERE")

model = client.model("facebook/detr-resnet-50")

model.load()

img_url = "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Cat_November_2010-1a.jpg/1200px-Cat_November_2010-1a.jpg"

result = model.run(img_url)

box_objects = result.get("output")

print(box_objects)

for box_object in box_objects:
    # Depending on the model, there may be additional props returned
    print(box_object)

    score = box_object["score"]
    label = box_object["label"]
    box = box_object["box"]

    print({"score": score, "label": label, "box": box})

# Fetch image and convert to base64
response = requests.get(img_url)
image_bytes = response.content

# Convert to PNG
image = Image.open(io.BytesIO(image_bytes)).convert("RGBA")
image_buffer_png = io.BytesIO()
image.save(image_buffer_png, format="PNG")
image_buffer_png.seek(0)


# Visualization function
def debug_image(src_img_buffer, box_objects):
    src_img = Image.open(src_img_buffer).convert("RGBA")
    draw = ImageDraw.Draw(src_img)
    colors = [
        (255, 0, 0, 128),  # Red
        (0, 255, 0, 128),  # Green
        (0, 0, 255, 128),  # Blue
        (255, 255, 0, 128),  # Yellow
    ]

    font = ImageFont.load_default()

    for index, box_object in enumerate(box_objects):
        box = box_object["box"]
        score = box_object["score"]
        label = box_object["label"]
        color = colors[index % len(colors)]

        # Draw the bounding box
        draw.rectangle(
            [(box["xmin"], box["ymin"]), (box["xmax"], box["ymax"])],
            outline=color,
            width=2,
        )

        # Prepare the label and score text
        text = f"{label} ({score:.2f})"

        # Get text size using textbbox method
        text_width, text_height = draw.textbbox((0, 0), text, font=font)[2:]

        # Draw a background rectangle for the text
        draw.rectangle(
            [
                (box["xmin"], box["ymin"] - text_height - 2),
                (box["xmin"] + text_width, box["ymin"]),
            ],
            fill=color,
        )
        draw.text(
            (box["xmin"], box["ymin"] - text_height - 2),
            text,
            fill=(255, 255, 255, 255),
            font=font,
        )

    # Save the image with boxes
    output_path = f"{WORKING_DIR}/testImage.png"
    src_img.save(output_path)


debug_image(image_buffer_png, box_objects)
